R/impact_central.R

```diff
@@ -299,3 +307,254 @@ get_birth_cohort <- function(data) {
+#' impact of vaccinating a cohort outside the cohort vaccinated (e.g.
+#' because of herd protection).
+#'
+#' @param baseline_impact Data frame of baseline impact data this needs to have
+#' columns country, burden_outcome, year, age, value
+#' @param focal_impact Data frame of focal impact data this needs to have
+#' columns country, burden_outcome, year, age, value
+#'
+#' @return Vaccine impact by country and birth year for burden outcomes as a
+#' data frame with columns country, year, burden_outcome and impact
+#' @export
+impact_by_birth_year <- function(baseline_impact, focal_impact) {
+  assert_has_columns(baseline_impact,
+                     c("country", "burden_outcome", "year", "age", "value"))
+  assert_has_columns(focal_impact,
+                     c("country", "burden_outcome", "year", "age", "value"))
+
+  baseline_impact$birth_cohort <- baseline_impact$year - baseline_impact$age
+  focal_impact$birth_cohort <- focal_impact$year - focal_impact$age
+  baseline <- stats::aggregate(value ~ birth_cohort + country + burden_outcome,
+                               baseline_impact,
+                               sum, na.rm = TRUE)
+  focal <- stats::aggregate(value ~ birth_cohort + country + burden_outcome,
+                            focal_impact,
+                            sum, na.rm = TRUE)
+  data <- merge(baseline, focal,
+                c("country", "burden_outcome", "birth_cohort"), sort = FALSE)
+  data$impact <- data$value.x - data$value.y
+  data[, c("country", "burden_outcome", "birth_cohort", "impact")]
+}
+
+#' Calculate impact by year of vaccination: activity type
+#'
+#' Impact by year of vaccination with impact ratio stratified by activity type.
+#' Stratifying impact ratio by activity type captures the differing effects of
+#' routine and campaign vaccination.
+#'
+#' To calculate impact by year of vaccination using impact ratios stratified by
+#' activity type, we assume that routine vaccination and campaign vaccination,
+#' which target multiple age groups, have different effects; for example due
+#' to dosage clustering. Hence, this method produces multiple,
+#' activity-specific impact ratios which we then multiply by the number of
+#' FVPs (fully vaccinated persons) to calculate impact.
+#'
+#' @param baseline_impact Data frame of baseline impact data this needs to have
+#' columns country, burden_outcome, vaccine_delivery, year, age, value
+#' @param focal_impact Data frame of focal impact data this needs to have
+#' columns country, burden_outcome, vaccine_delivery, year, age, value
+#' @param fvps Data frame of FVPs (fully vaccinated persons) needs to have
+#' columns country, year, activity_type, fvps other columns can be included
+#' and will be aggregated over.
+#' @param vaccination_years Years of vaccination of interest.
+#'
+#' @return Vaccine impact by country, activity type (routine or campaign), year
+#' and burden outcome
+#' @export
+impact_by_year_of_vaccination_activity_type <- function(
```

Comment:
everything looks good in this .R except this one. The reason being that you need different inputs to calculate the various impact ratios: for example impact_ratio_routine would need baseline impact = no vax and focal impact = routine scenario whereas the impact ratio for campaign would need baseline impact = routine and focal impact = routine+campaign. I think this is almost there just maybe worth clarifying this somehow. Of course then there is the case where you may only have routine or only have campaign so if there were 3 inputs you may have that one of them is NULL

Reply:
Am I understanding this correctly? Users will have 3 files? One for no vaccination scenario, one for routine and one for routine + campaign? (Plus file containing FVP data still) I can change this to work with 3 files instead. In that case should I change the terminology here and in other functions from `baseline` and `focal` scenarios to `no_vaccination` and `routine_campaign` or something similar? `baseline`, `focal_routine`, `focal_campaign`. And I guess in that case it won't have a column for `activity_type` this will be implied by file being used

Reply:
I think keeping baseline and focal works well as this links with the paper, we also have the situation (for cholera) where the focal scenario may be campaign only so it covers all bases. In terms of this particular function, allowing it work with 3 files would be ideal.

Reply:
Users may have 2 burden files for routine only diseases (novax and routine), 3 burden files for diseases with one routine and campaign (novax, routine & routine + campaign, or campaign & campaign_routine depending on how vaccine impact is understood), and so forth.

In general, ```focal_vaccine_impact_ratio = (baseline_burden - focal_burden) / (baseline_fvps - focal_fvps)```
so users need two burden files plus an fvp data file per vaccine of interest. 

BTW, routine impact is not necessarily novax v.s. routine. It can also be campaign v.s. routine+campaign. For example, HPV and MenA, which usually start with mass campaign and followed by routine introduction.